[{"text":"actually we can do more than just print out an error number","width":384},{"text":"there's a function that allows us to go from that error number to a useful string","width":256},{"text":"so we've got strerror and we can pass in a number","width":384},{"text":"like twelve [gibberish] I actually want the number that the system has set for us, so let's print that out as well and see what we get","width":512},{"text":"okay right here we go so yeah our read returned -1 to say hey I've failed","width":896},{"text":"we've got an error number that's now set to a non zero value","width":192},{"text":"and we've got bad file descriptor, which now makes sense to us","width":256},{"text":"yes we've passed in an invalid file descriptor number","width":192},{"text":"so before we've played around with things like printerror","width":320},{"text":"and it will print out something like","width":256},{"text":"It is a Friday let us go okay","width":576},{"text":"and perror takes the text we've got followed by okay here we go","width":832},{"text":"It's a friday let's go with a colon and guess what, the same message: bad file descriptor","width":448},{"text":"which is if you were a user a completely useless message","width":192},{"text":"bad file descriptor? sounds like the beginning of a bad movie","width":256},{"text":"no it's worse, so okay put this into your log file but don't display it to the user of course","width":512},{"text":"but now you can see how perror might actually be implemented in terms of this strerror function","width":512},{"text":"okay so that's our error number, it's set when a system call fails","width":512},{"text":"so if we go back to our little handout here","width":256},{"text":"so what is errno? it's just an int","width":512},{"text":"and when is it set?","width":64},{"text":"well when a system call fails","width":192},{"text":"you can find out what went wrong","width":768},{"text":"so nine is not a particularly... yes","width":768},{"text":"[question]","width":1024},{"text":"you mean because printerror adds a newline as well?","width":384},{"text":"is that your question?","width":128},{"text":"okay right so we don't need to deal with these numbers like nine, we don't need to remember that nine means bad file descriptor","width":704},{"text":"because guess what? on the man page of read","width":128},{"text":"[something] we discover the return value","width":192},{"text":"we got a list of errors, here they all are","width":384},{"text":"so lots of things can go wrong and if we wanted to write a robust program","width":704},{"text":"we might actually choose to handle some of these","width":256},{"text":"you might say okay in the generic sense, just crash the program or write to the log file","width":384},{"text":"but there might be others where you might decide to say print a message to the user","width":192},{"text":"and that's how we can do it, we can check errorno to see exactly why the system call failed","width":448},{"text":"and then we can just compare it to these constants","width":512},{"text":"so in our code here, we could say if errorno equals bad file descriptor EBADF, your file descriptor, your program is a mess!","width":1472},{"text":"okay so we are going to handle this particular problem","width":320},{"text":"what are we going to do about multiple threads?","width":960},{"text":"let's say I've got threads making system calls, they are calling mutex lock, they are calling sem wait, they are calling read, they are calling write","width":512},{"text":"if they fail, don't we have a problem?","width":320},{"text":"how would you know which one actually failed?","width":448},{"text":"suppose you're trying to read error just as another thread also failed","width":320},{"text":"houston we have a problem","width":256},{"text":"we don't know which thread set errorno","width":320},{"text":"muahahah yes, okay, errorno is not just an int","width":832},{"text":"each thread gets its own errorno integer","width":448},{"text":"how does that happen you ask? easy magic","width":960},{"text":"actually it uses something called thread local storage but never mind about that","width":512},{"text":"so that means if you want to know what happened inside your thread, easy, write errno","width":448},{"text":"if you want to know what error happened inside another thread, you can't","width":192},{"text":"you'd have to write the mechanics to share that information","width":256},{"text":"right, so it's actually local to each thread which is particularly clever","width":384},{"text":"yes, you can which brings us to the next question thank you very much for playing, right","width":448},{"text":"when is errno set to zero?","width":64},{"text":"when your process starts and that's it","width":256},{"text":"so here's a gotcha, that if we did something like if errorno is equal to something like we have the EBADF","width":1344},{"text":"then we just assume that that happened because of the previous call we just made","width":448},{"text":"that may not be true, perhaps this happened, perhaps errno was set many many many microseconds earlier in our program","width":640},{"text":"so errno is never reset back to zero","width":256},{"text":"if you want to set it back to zero, do it yourself","width":256},{"text":"and in fact that's quite a good kind of programming practice","width":256},{"text":"that if you intend to check its value","width":192},{"text":"then remember to set it to zero beforehand","width":2112}]